### ISSUE TITLE
Unexplained all-gather in batched array shuffling

### ISSUE THREAD

**Issue by hr0nix** [google/jax#20381]
### Description

I have an array, which is fully sharded along the first dimension. I'm trying to shuffle each row of the array using the code below. From what I understand, it should be an embarrassingly parallel operation, but XLA decides to insert an all-gather in the end of it for some reason that I cannot understand.

Two questions:
* What is this all-gather doing there?
* Is there a way to achieve what I want without any communications?

```
import jax
import jax.numpy as jnp
import numpy as np

def func(rng, inputs):
    batch_size, _, _ = inputs.shape

    def _f(rng, input):
        return jax.random.permutation(rng, input)
    
    rng = jax.random.split(rng, batch_size)
    return jax.vmap(_f)(rng, inputs)
    

def main():
    mesh = jax.sharding.Mesh(np.array(jax.devices()), ('dp'))
    rng_sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec())
    input_sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec('dp', None, None))
    f = jax.jit(
        func,
        in_shardings=(rng_sharding, input_sharding),
        out_shardings=(input_sharding),
    )

    inputs = jnp.zeros((24, 4096, 16))
    inputs = jax.device_put(inputs, input_sharding)
    rng = jax.random.PRNGKey(0)
    rng = jax.device_put(rng, rng_sharding)
    f = f.lower(rng, inputs).compile()
    text = f.as_text()
    print(text)
    if "all-gather-start" in text:
        print("All gather found")
    else:
        print("All gather not found")



if __name__ == '__main__':
    main()
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.24.3
python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
jax.devices (8 total, 8 local): [cuda(id=0) cuda(id=1) ... cuda(id=6) cuda(id=7)]
process_count: 1
platform: uname_result(system='Linux', node='hr0nix-dev-8-gpu', release='5.4.0-155-generic', version='#172-Ubuntu SMP Fri Jul 7 16:10:02 UTC 2023', machine='x86_64')
```

**Comment by hr0nix** [comment]
Ok, I was able to work around it with shard_map:

```
def main():
    mesh = jax.sharding.Mesh(np.array(jax.devices()), ('dp'))
    rng_pspec = jax.sharding.PartitionSpec()
    rng_sharding = jax.sharding.NamedSharding(mesh, rng_pspec)
    input_pspec = jax.sharding.PartitionSpec('dp', None, None)
    input_sharding = jax.sharding.NamedSharding(mesh, input_pspec)
    f = shard_map(
        func,
        mesh=mesh,
        in_specs=(rng_pspec, input_pspec),
        out_specs=input_pspec
    )
    f = jax.jit(
        f,
        in_shardings=(rng_sharding, input_sharding),
        out_shardings=input_sharding,
    )

    inputs = jnp.zeros((24, 4096, 16))
    inputs = jax.device_put(inputs, input_sharding)
    rng = jax.random.PRNGKey(0)
    rng = jax.device_put(rng, rng_sharding)
    f = f.lower(rng, inputs).compile()
    text = f.as_text()
    print(text)
    if "all-gather-start" in text:
        print("All gather found")
    else:
        print("All gather not found")

    print(f(rng, inputs).shape)
```

**Comment by hr0nix** [comment]
Ok, here is a fun fact: if I replace "jax.random.permutation" with "jax.lax.top_k" and leave `shard_map`, this code crashes with `Floating point exception` from native code (presumably XLA). Without shardmap it works fine.
